discovery/client.go

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2024 Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package discovery
+
+import (
+	"context"
+	"fmt"
+	"github.com/nuts-foundation/nuts-node/discovery/api/v1/client"
+	"github.com/nuts-foundation/nuts-node/discovery/log"
+	"time"
+)
+
+// clientUpdater is responsible for updating the presentations for the given services, at the given interval.
+// Callers should only call update().
+type clientUpdater struct {
+	services map[string]ServiceDefinition
+	store    *sqlStore
+	client   client.HTTPClient
+	verifier registrationVerifier
+}
+
+func newClientUpdater(services map[string]ServiceDefinition, store *sqlStore, verifier registrationVerifier, client client.HTTPClient) *clientUpdater {
+	return &clientUpdater{
+		services: services,
+		store:    store,
+		client:   client,
+		verifier: verifier,
+	}
+}
+
+// update starts a blocking loop that updates the presentations for the given services, at the given interval.
+// It returns when the context is cancelled.
+func (u *clientUpdater) update(ctx context.Context, interval time.Duration) {
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+	u.doUpdate(ctx)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			u.doUpdate(ctx)
+		}
+	}
+}
+
+func (u *clientUpdater) doUpdate(ctx context.Context) {
+	for _, service := range u.services {
+		if err := u.updateService(ctx, service); err != nil {
+			log.Logger().Errorf("Failed to update service (id=%s): %s", service.ID, err)
+		}
+	}
+}
+
+func (u *clientUpdater) updateService(ctx context.Context, service ServiceDefinition) error {
+	currentTag, err := u.store.getTag(service.ID)
+	if err != nil {
+		return err
+	}
+	// convert *Tag to *string
+	var currentTagStr *string
+	if !currentTag.Empty() {
+		currentTagStr = new(string)
+		*currentTagStr = string(currentTag)
+	}
+	presentations, newTag, err := u.client.Get(ctx, service.Endpoint, currentTagStr)
+	if err != nil {
+		return fmt.Errorf("failed to get presentations from discovery service (id=%s): %w", service.ID, err)
+	}
+	// *string back to *Tag
+	newTagStr := new(Tag)
+	*newTagStr = Tag(*newTag)
```

Comment:
this is again a lot of code to use an alias of a string. `client.Get` always returns a tag if err == nil (its even assumed that newTag != nil in the conversion here since that would result in a nil pointer panic), and the tag is not optional in the OAPI spec, so why covert to an optional/pointer value later on?